### ISSUE TITLE
Issue with path in windows

### ISSUE THREAD

**Issue by lguerard** [imcf/python-imcflibs#10]
The `parse_path` functions sometimes creates issues on windows when parsing a direct path copied.

* Using the script parameter for file `"U:\VAMP\guerard\LS7\new copy\SmalltileandZ - Copy.czi"` returns `'path':'U:\\VAMP\\guerard\\LS7\\new copy\\'` 
* Directly using `parse_path("U:\VAMP\guerard\LS7\new copy\SmalltileandZ - Copy.czi")` gives `'path': 'U:\\VAMP\\guerard\\LS7\new copy\\'` which then later can cause issues

**Comment by lguerard** [comment]
Adding `path = path.lower().replace("\n",r"\\n")` at the beggining of the script fixes it, but it doesn't feel like the proper way to do it... Let me know if you have any ideas @ehrenfeu

**Comment by ehrenfeu** [comment]
Sorry, I need to see the two results side by side:

```
           "U:\VAMP\guerard\LS7\new copy\SmalltileandZ - Copy.czi"
parse_path("U:\VAMP\guerard\LS7\new copy\SmalltileandZ - Copy.czi"
```

```
'path':'U:\\VAMP\\guerard\\LS7\\new copy\\'
'path': 'U:\\VAMP\\guerard\\LS7\new copy\\'
```

Now I'm confused. Could you point out or explain differently what the issue is exactly? The only thing I'm seeing is the space after the colon in the second section, but I guess that's just a copy-paste thing, no?

**Comment by ehrenfeu** [comment]
Ah, I see the `\n` in the second path now...

@imagejan  is probably right, calling it like this should work - the question is then rather if we can enforce this from within the function (as in, avoiding the string to be interpreted wrongly), or if that's already too late.

➡️ I'll need to look into this!

**Comment by rohangirishrao** [comment]
I guess replacing the `\n` back to "proper" alphabet at the top of the method should work, right? Converting to a raw string inside the function would be too late.
 `path = path.replace("\n", "\\n").replace("\t", "\\t")` ?

**Comment by ehrenfeu** [comment]
@lguerard could you create an additional test for this (preferred), or modify the existing one here:

**Comment by ehrenfeu** [comment]
@lguerard sorry if that's a stupid question, but is this issue supposed to be fixed by the changes from the PR's mentioned above (#24 & #26)?

**Comment by lguerard** [comment]
It's been a while, but I believe so yes.

**Comment by ehrenfeu** [comment]
Okay, just had another look at the `test_parse_path_windows_tabs_and_lines()` test and I think we should be fine closing this issue.

🐛 🔨 ✅ 

https://github.com/imcf/python-imcflibs/blob/2a124ba34abeb1c33140801a47d44ae8889aaa4f/tests/test_pathtools.py#L46-L51

**Comment by imagejan** [comment]
So what about:
```
parse_path("C:\bold\right\funny\valid\all\123\xab")
```
?

Here's what potentially _could_ go wrong:

```
\n - Newline
\t - Horizontal tab
\r - Carriage return
\b - Backspace
\f - Form feed
\v - Vertical tab
\' - Single quote
\" - Double quote
\\ - Backslash
\a - ASCII bell (BEL)
\N{name} - Unicode character by name (e.g., \N{GREEK CAPITAL LETTER DELTA})
\uxxxx - Unicode character with 16-bit hex value xxxx
\Uxxxxxxxx - Unicode character with 32-bit hex value xxxxxxxx
\ooo - Character with octal value ooo (up to 3 digits)
\xhh - Character with hex value hh (2 digits)
```

(**Disclaimer**: I didn't test any of this with an up-to-date python-imcflibs. Just posting here as a warning, as I think it would be better educating users about the right thing to do: using raw strings where appropriate.)

**Comment by ehrenfeu** [comment]
Actually, not true. The fix is still in a non-merged branch called [`pathtools-tab-newline`](https://github.com/imcf/python-imcflibs/commit/bcd0ca1728fd37a828cceb6f5dc22c1e33715cbe) as #26 was closed 🛑 🔀 unmerged (too many unrelated things, see https://github.com/imcf/python-imcflibs/issues/26#issuecomment-2476271692 for details).

**Comment by ehrenfeu** [comment]
And yes, @imagejan is obviously right. Requires more tests 🧪 .

**Comment by ehrenfeu** [comment]
Actually...

Since the entire library only really works within Fiji, I think it's enough to cover those use cases.

Therefore, I performed a few tests there, using the last `imcflibs` release (1.4) using this script:

```Python
#@ File mypath
#@ Integer myval

# INPUT for "mypath":
# X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\bunny

from imcflibs.pathtools import parse_path


def test(path):
    parsed = parse_path(path)
    res = "path: '%s'" % parsed["path"]
    res = "%s, basename: '%s'" % (res, parsed["basename"])
    print(res)


print("ORIG: '%s'" % mypath)
test(mypath)
test(str(mypath))
raw = r"X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\bunny"
test(raw)
cooked = "X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\bunny"
test(cooked)

print("\n\n")
```

On Linux 🐧 this works perfectly fine (except for the `cooked` string, which is expected), the output is:

```
ORIG: 'X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\bunny'
path: 'X:/violent/albeit/really/funny/007/u2324/x0e/new/turbo/', basename: 'bunny'
path: 'X:/violent/albeit/really/funny/007/u2324/x0e/new/turbo/', basename: 'bunny'
path: 'X:/violent/albeit/really/funny/007/u2324/x0e/new/turbo/', basename: 'bunny'
path: 'X:iolentlbeit
eallyunny/', basename: 'u2324
ew	urbounny'
```

Still need to test on :window: Windows though.

### Now, one question remains... 🤔

➡  What is the use-case of **directly** calling `parse_path()`  with a (hardcoded?) **non-raw** string?

https://github.com/imcf/python-imcflibs/blob/bcd0ca1728fd37a828cceb6f5dc22c1e33715cbe/src/imcflibs/pathtools.py#L12

**Comment by ehrenfeu** [comment]
Here's the output on :window: Windows:

```text
ORIG: 'X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\bunny'
path: 'X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\', basename: 'bunny'
path: 'X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\', basename: 'bunny'
path: 'X:\violent\albeit\really\funny\007\u2324\x0e\new\turbo\', basename: 'bunny'
path: 'X:iolentlbeit
eallyunny\', basename: 'u2324
ew	urbounny'
```

So from my current perspective, we should be all set... ✅

Any thoughts? 🤔

**Comment by ehrenfeu** [comment]
Note: if this really needs more treatment, we should look into `str.encode("raw_unicode_escape")`:

- [`str.encode()` (Built-in Types)][1]
- [codecs - Python Specific Encodings][2]

[1]: https://docs.python.org/2.7/library/stdtypes.html#str.encode
[2]: https://docs.python.org/2.7/library/codecs.html#python-specific-encodings

**Comment by imagejan** [comment]
The SciJava widgets handle string inputs correctly (i.e. entering `C:\test\new\path` will give you a string with correctly escaped backslashes `C:\\test\\new\\path`).

You're only screwed if you paste a Windows path directly into the script editor and forget using a raw `r""` string.

**Comment by imagejan** [comment]
> In my opinion we don't necessarily have to guard against this incorrect usage

I fully agree, and find it dangerous to do so (because it trains people wrong habits that will fail anywhere else).

I just can't see any inconsistent behavior on the Jython/Java side here...

**Comment by ehrenfeu** [comment]
Not here. But I found some when fiddling around with CPython 2.7 vs. Jython (vs. CPython 3.x) today.

Don't remember exactly any more though (my 🧠 refuses...), and it wouldn't change anything for this issue, it just made "debugging" even more enjoyable 😜

**Comment by ehrenfeu** [comment]
Closed with PR #42
